mavros/src/plugins/imu.cpp

```diff
@@ -0,0 +1,530 @@
+		imu_ned_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_ned_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		/** Store attitude in base_link ENU
+		 *  @snippet src/plugins/imu.cpp store_enu
+		 */
+		// [store_enu]
+		m_uas->update_attitude_imu_enu(imu_enu_msg);
+		// [store_enu]
+
+		/** Store attitude in aircraft NED
+		 *  @snippet src/plugins/imu.cpp store_ned
+		 */
+		// [store_enu]
+		m_uas->update_attitude_imu_ned(imu_ned_msg);
+		// [store_ned]
+
+		/** Publish only base_link ENU message
+		 *  @snippet src/plugins/imu.cpp pub_enu
+		 */
+		// [pub_enu]
+		imu_pub.publish(imu_enu_msg);
+		// [pub_enu]
+	}
+
+	/**
+	 * @brief fill and publish IMU data_raw message; store linear acceleration for IMU data
+	 * @param[in] header		Message frame_id and timestamp
+	 * @param[in,out] gyro		Orientation in the base_link ENU frame
+	 * @param[in,out] accel_enu	Linear acceleration in the base_link ENU frame
+	 * @param[in,out] accel_ned	Linear acceleration in the aircraft NED frame
+	 */
+	void publish_imu_data_raw(std_msgs::Header &header, Eigen::Vector3d &gyro,
+				Eigen::Vector3d &accel_enu, Eigen::Vector3d &accel_ned)
+	{
+		auto imu_msg = boost::make_shared<sensor_msgs::Imu>();
+
+		// Fill message header
+		imu_msg->header = header;
+
+		tf::vectorEigenToMsg(gyro, imu_msg->angular_velocity);
+		tf::vectorEigenToMsg(accel_enu, imu_msg->linear_acceleration);
+
+		// Save readings
+		linear_accel_vec_enu = accel_enu;
+		linear_accel_vec_ned = accel_ned;
+
+		imu_msg->orientation_covariance = unk_orientation_cov;
+		imu_msg->angular_velocity_covariance = angular_velocity_cov;
+		imu_msg->linear_acceleration_covariance = linear_acceleration_cov;
+
+		// Publish message [ENU frame]
+		imu_raw_pub.publish(imu_msg);
+	}
+
+	/**
+	 * @brief Publish magnetic field data
+	 * @param[in] header		Message frame_id and timestamp
+	 * @param[in,out] mag_field	Magnetic field in the base_link ENU frame
+	 */
+	void publish_mag(std_msgs::Header &header, Eigen::Vector3d &mag_field)
+	{
+		auto magn_msg = boost::make_shared<sensor_msgs::MagneticField>();
+
+		// Fill message header
+		magn_msg->header = header;
+
+		tf::vectorEigenToMsg(mag_field, magn_msg->magnetic_field);
+		magn_msg->magnetic_field_covariance = magnetic_cov;
+
+		// Publish message [ENU frame]
+		magn_pub.publish(magn_msg);
+	}
+
+	/* -*- message handlers -*- */
+
+	/**
+	 * @brief Handle ATTITUDE MAVlink message.
+	 * Message specification: http://mavlink.org/messages/common/#ATTITUDE
+	 * @param[in] msg	Received Mavlink msg
+	 * @param[in,out] att	ATTITUDE msg
+	 */
+	void handle_attitude(const mavlink::mavlink_message_t *msg, mavlink::common::msg::ATTITUDE &att)
+	{
+		if (has_att_quat)
+			return;
+
+		/** Orientation on the NED-aicraft frame:
+		 *  @snippet src/plugins/imu.cpp ned_aircraft_orient1
+		 */
+		// [ned_aircraft_orient1]
+		auto ned_aircraft_orientation = ftf::quaternion_from_rpy(att.roll, att.pitch, att.yaw);
+		// [ned_aircraft_orient1]
```

Comment:
What is that? Also i don't think that we need function body comments in doxygen.

Reply:
This is how you include a code snippet on Doxygen docs. Without it, the comments above the code lines would appear loose on the documentation (just the lines, and not the code associated to the comment). That way you guarantee that the code you are referring to on the comment is also documented on the output of Doxygen.